### ISSUE TITLE
Submit CFP to RubyKaigi Takeout 2021

### ISSUE THREAD

**Issue by fujimotos** [fluent/fluentd#3435]
**Problem Description**

* [RubyKaigi Takeout 2021](https://rubykaigi.org/2021-takeout) is calling for speakers.
* We are palnning to apply for the conference as a speaker.
*  For that purpose, we need to submit a CFP before the proposal deadline (2021/6/30 23:59).

**Assginee**

Me (fujimotos)

**Timeline**

* I will write a CFP draft and  post the draft to this ticket next Monday.
* I'm thinking to talk something like "Fluentd: Ship a Ruby Appliction to Windows".
  * td-agent-builder, Windows event support etc.
  * Basically the stuffs we've done for Windows last months.
* Review & revise the draft, and submit it to rubykaigi.org on Tue/Wed.

**Comment by fujimotos** [comment]
@ashie @kenhys @kou This is the CFP draft for Rubykaigi Takeout 2021.

Please comment here if you have any suggestions (preferably by Tuesday
as the submission close is next Wed).

```markdown
# Title

Fluentd: How to ship a Ruby Application to Windows

# Abstract

Shipping a Ruby application to Windows is a challenging task. For starters, we
cannot even expect that Ruby is handily pre-installed there!

We (Fluentd team) maintain a set of Windows packages to make it easier to
deploy and run Fluentd on Windows. In this talk, we discuss the general problems
of shipping Ruby applications to Windows, and explain how we addressed them.

# Details

The goal of my talk is to let attendants obtain the general knowledge of how to
distribute Ruby applications on Windows.

Here is the (planned) outline of my talk:

* What is Fluentd? (intro)
* Windows and Ruby
  * Myth: "Ruby is cross-platform. My app should just work fine!"
  * Not all platform incosistencies are abstracted away.
  * Not all gems are Windows-ready.
* Detailed know-hows
  * Porting your app to Windows.
  * How to create a Windows installer (WIX/NSIS...)
  * How to include a copy of Ruby in your package.
  * How to manage gems for your app on Windows.
  * How to run your Windows app as Windows Service.
  * ...
* Open problems
  * Unix culture vs Windows culture
* Conclusion

# Pitch

Ruby developers demographic is heavily leaned toward Unix (Linux/BSD/macOS).
Generally speaking, there is much less community knowledge and expertise
regarding Windows.

However Windows is the most widely used desktop operating system, especially
in the business space. There are indeed a lot of demands for Ruby apps to support
Windows. I see here a certain gap waiting to be filled between Ruby devs (Unix)
and general user demographic (mostly Windows).

Fluentd project is maintaining the Windows distribution for years, and I think
we can provide a unique perspective on this topic.
```

**Comment by kenhys** [comment]
One more topic for "Detailed know-hows"
* How to apply a fix for a copy of Ruby (and gems)

**Comment by kou** [comment]
RubyKaigi is a deeply technical conference for Ruby itself.

http://blade.nagaokaut.ac.jp/cgi-bin/scat.rb/ruby/ruby-core/104408

> we're looking for hardcore tech talks from Ruby hackers!

It seems that this talk focuses on packaging mainly. Can we focus on deeply technical topic related to Ruby as the main topic?

**Comment by ashie** [comment]
I'm not sure they are suitable for RubyKaigi though, following topics are Ruby on Windows related things in Fluentd I know.

* How to treat Windows spefic problems
  * Cannot use `WSAGetLastError` on Ruby 3.0: https://github.com/ruby/fiddle/issues/72, https://github.com/treasure-data/serverengine/pull/112
  * Cannot use UNIX signals, we replace them with [pipe communication](https://github.com/treasure-data/serverengine/blob/master/lib/serverengine/command_sender.rb) or [Win32 Event](https://github.com/fluent/fluentd/blob/da162e15bc95b1263f92eb0622c714f8b9d303b9/lib/fluent/supervisor.rb#L209)
  * System error code: errno vs GetLastError (can be mapped by [`rb_w32_map_errno`](https://github.com/treasure-data/serverengine/blob/08193f12e8a36518a35244b94f2865d974538450/lib/serverengine/winsock.rb#L112-L117))
  * [winsock](https://github.com/treasure-data/serverengine/blob/master/lib/serverengine/winsock.rb)
    * linger related issues: #1968 #2398 #2868 #2959
  * [file handle](https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/file_wrapper.rb)
* How to get Windows EventLog (on Ruby)
* How to use TLS certifications in Windows Certificate Store (on Ruby)
* How to run as Windows Service (on Ruby)

**Comment by kou** [comment]
> * How to get Windows EventLog (on Ruby)
> * How to use TLS certifications in Windows Certificate Store (on Ruby)
> * How to run as Windows Service (on Ruby)

Could you provide links of their implementations?

**Comment by ashie** [comment]
> Could you provide links of their implementations?

I've added them to the comment.

**Comment by kou** [comment]
Thanks.
I think that describing how to implement XXXs on Ruby deeply is more suitable topic than packaging for RubyKaigi's talk.

**Comment by fujimotos** [comment]
@kou @ashie I created a new CFP draft. The presentation topic is Ractor and Fluentd.

* I'm thinking to add a Ractor-enabled version of the following function:
  * https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/parser_ltsv.rb#L38-L48
  * https://github.com/fluent/fluentd/blob/master/lib/fluent/plugin/in_tail.rb#L601-L607
* Add an experimental support and take some benchmark. 
* LTSV parsing should be fairly easily Ractor-able.

**RubyKaigi Takout 201 CFP (draft v2)**

```markdown
# Title

Speed up data processing of Fluentd using Ractor

# Abstract

Prallelization is becoming the norm in moden programming. As more cores are added to CPUs, programs are expected to utilize these additional cores.

Ractor is the new parallelism mechanism available since Ruby 3.0. Recently, Fluentd team experimented with this relatively-new feature to make the text parser faster. In this talk, we introduce the general concept of Ractor and discuss our experience with it.

# Details

The goal of my talk is to let attendants obtain the general knowledge of Ractor, the bleeding-edge multi-thereading feature

Here is the (planned) outline of my talk:

* What is Ractor? (Introduction)
  * Ractor = safe threading with easier interface
  * Feature: Parallel processing (with multiple Ractors)
  * Feature: Message passing
  * Restriction: Cannot share data
  * Restriction: Most gems do not support it (yet)
* Make Fluentd faster to parse texts using Ractor
  * Basic: How Fluentd parser works
  * Making LTSV parser faster (1)
  * Making LTSV parser faster (2)
  * Making LTSV parser faster (3)
  * Experiment
  * Result 
* Open issues
  * Ractor: still "experiment" feature
  * We need to make more gems Ractor-ready
* Conclusion

# Pitch

Parallelism becomes a very hot topic recently. As CPU vendors add more and more
cores to their products, more programing languages are starting to give the special
consideration to parallel processing.

Ruby was historically not very good at concurrent programming (e.g. GIL issue).
However, the status has rapidly changed in the last few years. Especially
the introduction of Ractor was a big factor.

In this talk, we're going to introduce Ractor to general audience and show how
it can be made used of to speed up data processing.
```

**Comment by fujimotos** [comment]
Here is a PoC implementation of Ractor-enabled `in_tail`.

https://github.com/fujimotos/fluentd/commit/7fdb403e894bb9825309b30e0f71d116eda4dee8

Using the following configuration:

```xml
<source>
  @type tail
  tag test.log
  path test.log
  read_from_head true
  <parse>
    @type ltsv
  </parse>
  <buffer>
    @type memory
    flush_interval 1s
  </buffer>
</source>

<match test.**>
  @type flowcounter_simple
  unit second
</match>
```

I can confirm that it actually utilizes 8 threads (cores):

![ractor2](https://user-images.githubusercontent.com/8974561/123599878-0db69a00-d831-11eb-83f7-420508121a93.gif)

A caveat is: actually the current implementation is a little slower
than the current single-threaded implementation. (Maybe due
to Ractor's overhead?)

But anyway, I think we can polish the PoC up and make a talk on it.

**Comment by kou** [comment]
It seems that this is a good starting point.
If we can improve it, we can talk about how to use Ractor for performance.

BTW, is it faster than `Thread` based implementation?

**Comment by kou** [comment]
Abound 30% CPU usage per thread may be low.
Each Ractor may be more idle than expected.

**Comment by fujimotos** [comment]
@kou  We don't have any thread implementation on in_tail.
`in_tail` basically uses a "one thread for one file" model.

Here is some performance stats:

| Mode | Speed [lines/s] |
| ---- | ---- |
| single-thread          | 238,934 |
| Ractor (2 workers) |  203,435 |
| Ractor (8 workers) |  211,627 |

Note that the test machine has 8 CPUs (Azure F8s with Xeon 8272CL)
Each CPU core seems to max out at around 30-60%. Never goes 100%.

**Comment by kou** [comment]
Is `in_tail` a CPU bound process?
If it's an IO bound process, it may not be suitable for Ractor.

**Comment by fujimotos** [comment]
I experimened with Ractor using another part of Fluentd last night.

https://github.com/fujimotos/fluentd/blob/aa5c1626/lib/fluent/plugin/filter_grep_ractor.rb#L81-L104

It didn't work. Ractor was actually slower than a single-thread implementation.
Other people seem to report similar results too:

* [More benchmarks, Ractors vs Forks + Redis + Marshal
](https://www.reddit.com/r/ruby/comments/jk9zvz/more_benchmarks_ractors_vs_forks_redis_marshal/)
* [Ractors in Ruby 3](https://speakerdeck.com/noahgibbs/ractors-in-ruby-3?slide=17)

So, apparently, Ractor seem not to be eligible for production usage (yet)...

**Comment by kou** [comment]
> Ractor seem not to be eligible for production usage (yet)...

Yes. Ractor is still an experimental feature.

**Comment by fujimotos** [comment]
**RubyKaigi Takout 201 CFP (draft v3)**

```markdown
# Title

Ractor: Experimenting with the new parallelization feature

# Abstract

Parallelization is becoming the norm in model programming. As more cores are
added to CPUs, programmers are expected to utilize these additional cores.

Ractor is the new parallelism mechanism available since Ruby 3.0. Recently,
Fluentd team experimented with this relatively-new feature to see if it can
accelarete data processing.

In this talk, we explain our experience with Ractor and discuss the current
status of this technology.

# Details

The goal of my talk is to let attendants obtain the general knowledge of Ractor,
the bleeding-edge multi-thereading model.

Here is the (planned) outline of my talk:

* What is Ractor? (Introduction)
  * Ractor = safe threading with easier interface
  * Feature: Parallel processing (with multiple Ractors)
  * Restriction: Data sharing
  * Restriction: Gem support
* Practical performance of Ractor
  * Basic: How Fluentd works
  * Make Fluentd Plugin Ractor-ready (1)
  * Make Fluentd Plugin Ractor-ready (2)
  * Make Fluentd Plugin Ractor-ready (3)
  * Experiment & Result 
* Open problems
* Conclusion

# Pitch

Parallelism becomes a hot topic recently. As multi-CPU architecture becomes
popular, programming languages are starting to give the special consideration
to parallel processing.

Ruby was historically not very good at concurrent programming (e.g. GIL vs Thread).
However, the status has rapidly changed in the last few years. Especially
the introduction of Ractor in Ruby 3 was deemed as a big step ahead.

In this presentation, we want to talk about this breeding-edge technology from
application developers' perspective
```

**Comment by fujimotos** [comment]
**RubyKaigi Taketout 2021 CFP draft (v4)**

```markdown
# Title

Ractor: Experimenting with the new parallelization approach

# Abstract

Parallelism has become the norm in modern programming. As more cores are
added to CPU chips, developers are expected to integrate multi-threading
into programs to utilize these additional cores.

Ractor is the new multi-threading mechanism introduced in Ruby 3.0.0.
Recently, Fluentd team experimented with this new technique to see if it
can bring higher data throughputs.

In this talk, we explain our experience with Ractor and discuss the
current state of the technology.

# Details

The goal of this talk is to show how to use Ractor, the breeding-edge
multithreading feature, with practical use cases.

Here is the (planned) outline of my talk:

* Introduction: What is Ractor?
  * Ractor = Actor-based multi-threading
  * Ractor can be executed in parallel (vs Threads)
  * Ractor is easier and safer to use
  * Caveats (1) Limited object sharing
  * Caveats (2) Not all gems support it yet
* Practical performance of Ractor
  * Basics: How Fluentd works
  * Make Fluentd Plugin Ractor-ready (1)
  * Make Fluentd Plugin Ractor-ready (2)
  * Make Fluentd Plugin Ractor-ready (3)
  * Experiment 
  * Result 
* Open problems
  * Some cases Ractor cannot handle well
  * On Gem ecosystem and Ractor
  * The experimental side of Ractor
* Conclusion

# Pitch

Parallelism is a hot topic right now. More and more programming languages
are integrating multi-threading technologies into their design (e.g. Go's
goroutines), and some are getting real traction.

Ruby was historically heavily restricted to a single thread. However, the
status has been rapidly changing in recent years. In particular, the
introduction of Ractor in 3.0.0 was regarded as a big step ahead.

Still, this technology is relatively new, and not much experience is
accumulated on community. I think our presentation will act as a good
introduction for general Ruby developers.

**Comment by fujimotos** [comment]
@kou @ashie @kenhys  CFP submitted at Jun 30 23:30.
The actual submitted version is https://github.com/fluent/fluentd/issues/3435#issuecomment-871794261.

I close this issue as done now.
